File: core/reader/src/main/kotlin/com/readler/core/reader/PdfReaderEngine.kt
package com.readler.core.reader

import com.readler.core.model.BookFormat
import com.readler.core.reader.locator.ReaderLocatorCodec
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File
import java.io.FileNotFoundException

class PdfReaderEngine : ReaderEngine {
    override val format: BookFormat = BookFormat.PDF

    override suspend fun open(source: ReaderSource): ReaderDocument = withContext(Dispatchers.IO) {
        val file = File(source.filePath)
        if (!file.exists()) {
            throw FileNotFoundException("PDF file missing: ${source.filePath}")
        }

        ReaderDocument.Pdf(
            filePath = source.filePath,
            initialLocator = source.initialLocator,
            initialPageIndex = ReaderLocatorCodec.decodePdfPageIndex(source.initialLocator) ?: 0
        )
    }
}

File: lib/core/reader/pdf_reader_engine.dart
import 'dart:io';

import '../model/book_format.dart';
import 'locator/reader_locator_codec.dart';
import 'reader_document.dart';
import 'reader_engine.dart';

class PdfReaderEngine implements ReaderEngine {
  @override
  BookFormat get format => BookFormat.pdf;

  @override
  Future<ReaderDocument> open({
    required String filePath,
    String? initialLocator,
  }) async {
    final file = File(filePath);
    if (!file.existsSync()) {
      throw FileSystemException('PDF file missing', filePath);
    }
    return PdfDocument(
      filePath: filePath,
      initialLocator: initialLocator,
      initialPageIndex:
          ReaderLocatorCodec.decodePdfPageIndex(initialLocator) ?? 0,
    );
  }
}
